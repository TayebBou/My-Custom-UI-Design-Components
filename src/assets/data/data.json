{
  "photos": [
    {
      "id": "AUDI_R8",
      "label": "Extérieur et intérieur",
      "photos": [
        {
          "url_full": "/images/AUDI_R8_1.webp",
          "url_vignette": "/images/AUDI_R8_1.webp"
        },
        {
          "url_full": "/images/AUDI_R8_2.webp",
          "url_vignette": "/images/AUDI_R8_2.webp"
        },
        {
          "url_full": "/images/AUDI_R8_3.webp",
          "url_vignette": "/images/AUDI_R8_3.webp"
        },
        {
          "url_full": "/images/AUDI_R8_4.webp",
          "url_vignette": "/images/AUDI_R8_4.webp"
        },
        {
          "url_full": "/images/AUDI_R8_5.webp",
          "url_vignette": "/images/AUDI_R8_5.webp"
        },
        {
          "url_full": "/images/AUDI_R8_6.webp",
          "url_vignette": "/images/AUDI_R8_6.webp"
        },
        {
          "url_full": "/images/AUDI_R8_7.webp",
          "url_vignette": "/images/AUDI_R8_7.webp"
        },
        {
          "url_full": "/images/AUDI_R8_8.webp",
          "url_vignette": "/images/AUDI_R8_8.webp"
        },
        {
          "url_full": "/images/AUDI_R8_9.webp",
          "url_vignette": "/images/AUDI_R8_9.webp"
        },
        {
          "url_full": "/images/AUDI_R8_10.webp",
          "url_vignette": "/images/AUDI_R8_10.webp"
        }
      ]
    },
    {
      "id": "MERCCC",
      "label": "Extérieur et intérieur",
      "photos": [
        {
          "url_full": "/images/MERCEDES_CLASSE_C_1.webp",
          "url_vignette": "/images/MERCEDES_CLASSE_C_1.webp"
        },
        {
          "url_full": "/images/MERCEDES_CLASSE_C_2.webp",
          "url_vignette": "/images/MERCEDES_CLASSE_C_2.webp"
        },
        {
          "url_full": "/images/MERCEDES_CLASSE_C_3.webp",
          "url_vignette": "/images/MERCEDES_CLASSE_C_3.webp"
        },
        {
          "url_full": "/images/MERCEDES_CLASSE_C_4.webp",
          "url_vignette": "/images/MERCEDES_CLASSE_C_4.webp"
        },
        {
          "url_full": "/images/MERCEDES_CLASSE_C_5.webp",
          "url_vignette": "/images/MERCEDES_CLASSE_C_5.webp"
        },
        {
          "url_full": "/images/MERCEDES_CLASSE_C_6.webp",
          "url_vignette": "/images/MERCEDES_CLASSE_C_6.webp"
        },
        {
          "url_full": "/images/MERCEDES_CLASSE_C_7.webp",
          "url_vignette": "/images/MERCEDES_CLASSE_C_7.webp"
        },
        {
          "url_full": "/images/MERCEDES_CLASSE_C_8.webp",
          "url_vignette": "/images/MERCEDES_CLASSE_C_8.webp"
        },
        {
          "url_full": "/images/MERCEDES_CLASSE_C_9.webp",
          "url_vignette": "/images/MERCEDES_CLASSE_C_9.webp"
        },
        {
          "url_full": "/images/MERCEDES_CLASSE_C_10.webp",
          "url_vignette": "/images/MERCEDES_CLASSE_C_10.webp"
        },
        {
          "url_full": "/images/MERCEDES_CLASSE_C_11.webp",
          "url_vignette": "/images/MERCEDES_CLASSE_C_11.webp"
        }
      ]
    }
  ]
}
