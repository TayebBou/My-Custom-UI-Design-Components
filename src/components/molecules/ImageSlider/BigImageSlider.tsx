import { FC, useState } from "react";
import { IPhoto } from '../../../shared/types/photo.model';
import styles from './BigImageSlider.module.css'

type SliderProps = {
    photos: IPhoto[]
}

const BigImageSlider:FC<SliderProps> = (props) => {

    const photos = props.photos[0].photos;

    const [x, setX] = useState(0)

    const handleRight = () => {
        if(x !== (-100 * (photos.length-1))){
            setX(x => x - 100)
        }
    }

    const handleLeft = () => {
        if(x !== 0){
            setX(x => x + 100)
        }
    }

    return (
        <div className={styles["full-images"]}>
            {photos.map((p) => (
              <img key={p.url_full} src={p.url_full} alt="car outside" style={{ maxWidth: "100%", transition: '0.5s', transform: `translateX(${x}%)`}} />
            ))}
            <button disabled={x === (-100 * (photos.length-1))} className={`${styles["slider-button"]} ${styles.right}`} onClick={handleRight}>{'>'}</button>
            <button disabled={x === 0} className={`${styles["slider-button"]} ${styles.left}`} onClick={handleLeft}>{'<'}</button>
          </div>
    )
}

export default BigImageSlider