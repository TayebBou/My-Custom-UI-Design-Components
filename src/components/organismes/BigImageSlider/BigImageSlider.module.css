.full-images {
  display: flex;
  overflow-x: hidden;
  max-width: 70vw;
  margin: auto;
  position: relative;
  height: 100%;
}

.fullscreen {
  max-width: 40em;
}

.slider-button {
  position: absolute;
  width: 3em;
  font-size: 2em;
  background-color: #00000000;
  height: -webkit-fill-available;
  color: white;
  z-index: 1;
}

.slider-button:enabled:hover {
  background-color: #00000042;
}

.right {
  right: 0;
}

.left {
  left: 0;
}

.span {
  position: absolute;
  font-size: 1em;
  background-color: #000000;
  color: white;
  bottom: 0;
  left: 0;
  padding: 0.3em 1em;
}

@media (orientation: portrait) and (display-mode: fullscreen) {
  .full-images {
    max-width: 100vw;
    margin: auto;
    position: relative;
    height: 40%;
    top: 50%;
    transform: translateY(-50%);
  }
}

@media (orientation: portrait) {
  .full-images {
    max-width: 100%;
  }
}