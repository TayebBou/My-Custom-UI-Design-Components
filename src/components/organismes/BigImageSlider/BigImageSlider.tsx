import { FC, useRef, useState } from 'react'
import { IPhoto } from '../../../shared/types/photo.model'
import Button from '../../atoms/Button/Button'
import FullScreen from '../../molecules/FullScreen/FullScreen'
import styles from './BigImageSlider.module.css'

type SliderProps = {
  photos: IPhoto[]
}

const BigImageSlider: FC<SliderProps> = (props) => {
  const photos = props.photos[0].photos

  const [x, setX] = useState(0)
  const images: React.LegacyRef<HTMLDivElement> = useRef(null)

  const handleRight = () => {
    if (x !== -100 * (photos.length - 1)) {
      setX((x) => x - 100)
    }
  }

  const handleLeft = () => {
    if (x !== 0) {
      setX((x) => x + 100)
    }
  }

  const openFullScreen = () => {
    const element = images.current

    if (element) {
      if (!document.fullscreenElement) {
        element.requestFullscreen()
      } else {
        if (document.exitFullscreen) {
          document.exitFullscreen()
        }
      }
    }
  }

  return (
    <div ref={images} className={styles.fullscreen}>
      <div className={styles['full-images']}>
        {photos.map((p) => (
          <img
            key={p.url_full}
            src={p.url_full}
            alt="car outside"
            style={{
              maxWidth: '100%',
              transition: '0.5s',
              transform: `translateX(${x}%)`,
            }}
          />
        ))}
        <Button
          disabled={x === -100 * (photos.length - 1)}
          className={`${styles['slider-button']} ${styles.right}`}
          onClick={handleRight}
        >
          {'>'}
        </Button>
        <Button
          disabled={x === 0}
          className={`${styles['slider-button']} ${styles.left}`}
          onClick={handleLeft}
        >
          {'<'}
        </Button>
        <span className={styles.span}>
          {Math.abs(x / 100) + 1} / {photos.length}
        </span>
        <FullScreen onClick={openFullScreen} />
      </div>
    </div>
  )
}

export default BigImageSlider
