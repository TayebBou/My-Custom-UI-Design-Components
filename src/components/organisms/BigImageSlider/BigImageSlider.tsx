import { FC, useEffect, useRef, useState } from "react";
import { IPhoto } from "../../../shared/types/photo.model";
import Button from "../../atoms/Button/Button";
import FullScreen from "../../molecules/FullScreen/FullScreen";
import styles from "./BigImageSlider.module.scss";

type BigImageSliderProps = {
  photos: IPhoto[];
};

const BigImageSlider: FC<BigImageSliderProps> = (props) => {
  const photos = props.photos[0].photos;

  const [x, setX] = useState(0);
  const [isRendered, setIsRendered] = useState(false);
  const images: React.LegacyRef<HTMLDivElement> = useRef(null);

  const handleRight = () => {
    if (
      photos.length &&
      photos.length > 1 &&
      x > -100 * (photos.length - 1) &&
      x <= 0 &&
      x % 100 === 0 &&
      Number.isInteger(x)
    ) {
      setX((x) => x - 100);
    }
  };

  const handleLeft = () => {
    if (
      photos.length &&
      photos.length > 1 &&
      x < 0 &&
      x >= -100 * (photos.length - 1) &&
      x % 100 === 0 &&
      Number.isInteger(x)
    ) {
      setX((x) => x + 100);
    }
  };

  useEffect(() => {
    setIsRendered(true);
  }, []);

  return (
    <div ref={images} className={styles.fullscreen}>
      <div className={styles["full-images"]}>
        {photos.map((p) => (
          <img
            key={p.url_full}
            src={p.url_full}
            alt="car outside"
            style={{
              maxWidth: "100%",
              transition: "0.5s",
              transform: `translateX(${x}%)`,
            }}
          />
        ))}
        <Button
          disabled={x === -100 * (photos.length - 1)}
          className={`${styles["slider-button"]} ${styles.right}`}
          onClick={handleRight}
        >
          {">"}
        </Button>
        <Button
          disabled={x === 0}
          className={`${styles["slider-button"]} ${styles.left}`}
          onClick={handleLeft}
        >
          {"<"}
        </Button>
        <span className={styles.span}>
          {Math.abs(x / 100) + 1} / {photos.length}
        </span>
        {isRendered ? (
          <FullScreen element={images.current as HTMLElement} />
        ) : (
          <FullScreen />
        )}
      </div>
    </div>
  );
};

export default BigImageSlider;
