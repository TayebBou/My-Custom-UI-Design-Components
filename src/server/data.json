{
  "photos": [
    {
      "id": "E40012456",
      "label": "Extérieur",
      "photos": [
        {
          "url_full": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/ibiza/ibiza1-full.jpeg",
          "url_vignette": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/ibiza/ibiza1-vignette.jpeg"
        },
        {
          "url_full": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/ibiza/ibiza2-full.jpeg",
          "url_vignette": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/ibiza/ibiza2-vignette.jpeg"
        },
        {
          "url_full": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/ibiza/ibiza3-full.jpeg",
          "url_vignette": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/ibiza/ibiza3-vignette.jpeg"
        }
      ]
    },
    {
      "id": "F40012456",
      "label": "Extérieur",
      "photos": [
        {
          "url_full": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/captur/captur1-full.jpeg",
          "url_vignette": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/captur/captur1-vignette.jpeg"
        },
        {
          "url_full": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/captur/captur2-full.jpeg",
          "url_vignette": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/captur/captur2-vignette.jpeg"
        },
        {
          "url_full": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/captur/captur3-full.jpeg",
          "url_vignette": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/captur/captur3-vignette.jpeg"
        }
      ]
    },
    {
      "id": "G40012456",
      "label": "Extérieur",
      "photos": [
        {
          "url_full": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/i8/i81-full.jpeg",
          "url_vignette": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/i8/i81-vignette.jpeg"
        },
        {
          "url_full": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/i8/i82-full.jpeg",
          "url_vignette": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/i8/i82-vignette.jpeg"
        },
        {
          "url_full": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/i8/i83-full.jpeg",
          "url_vignette": "https://glc-codingtest.s3.eu-west-1.amazonaws.com/img-vo/i8/i83-vignette.jpeg"
        }
      ]
    }
  ]
}
