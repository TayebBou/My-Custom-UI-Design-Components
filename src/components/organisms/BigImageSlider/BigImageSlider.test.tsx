import { screen, render, fireEvent, waitFor } from "@testing-library/react";
import data from "../../../assets/data/data.json";
import BigImageSlider from "./BigImageSlider";

const props = {
  photos: data.photos,
};

describe("BigImageSlider component", () => {
  test("should render correctly with props", () => {
    // Arrange
    const { asFragment } = render(<BigImageSlider {...props} />);
    // Assert
    expect(asFragment()).toMatchSnapshot();
  });

  test("when user click on right button the slide go to the next image", async () => {
    // Arrange
    render(<BigImageSlider {...props} />);
    const BigImages = await screen.findAllByAltText("car outside", {
      exact: true,
    });
    expect(BigImages[0]).toHaveStyle({ transform: `translateX(0%)` });
    // Right Button of Big Image Slider
    const RightButton = screen.getByText(">", { exact: true });
    // Act
    fireEvent.click(RightButton);
    // Assert
    await waitFor(() => {
      expect(BigImages[0]).toHaveStyle({ transform: `translateX(-100%)` });
    });
  });
  test("when user click on left button the slide go back to the previous image", async () => {
    // Arrange
    render(<BigImageSlider {...props} />);
    const BigImages = await screen.findAllByAltText("car outside", {
      exact: true,
    });
    expect(BigImages[0]).toHaveStyle({ transform: `translateX(0%)` });
    // slide two times and back one time
    const RightButton = screen.getByText(">", { exact: true });
    // Act
    fireEvent.click(RightButton);
    fireEvent.click(RightButton);
    const LeftButton = screen.getByText("<", { exact: true });
    fireEvent.click(LeftButton);
    // Assert
    await waitFor(() => {
      expect(BigImages[0]).toHaveStyle({ transform: `translateX(-100%)` });
    });
  });
});
