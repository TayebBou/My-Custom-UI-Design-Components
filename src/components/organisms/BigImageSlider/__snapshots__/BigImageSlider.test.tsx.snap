// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BigImageSlider component should render correctly with props 1`] = `
<DocumentFragment>
  <div
    class="fullscreen"
  >
    <div
      class="full-images"
    >
      <img
        alt="car outside"
        src="/images/AUDI_R8_1.webp"
        style="max-width: 100%; transition: 0.5s; transform: translateX(0%);"
      />
      <img
        alt="car outside"
        src="/images/AUDI_R8_2.webp"
        style="max-width: 100%; transition: 0.5s; transform: translateX(0%);"
      />
      <img
        alt="car outside"
        src="/images/AUDI_R8_3.webp"
        style="max-width: 100%; transition: 0.5s; transform: translateX(0%);"
      />
      <img
        alt="car outside"
        src="/images/AUDI_R8_4.webp"
        style="max-width: 100%; transition: 0.5s; transform: translateX(0%);"
      />
      <img
        alt="car outside"
        src="/images/AUDI_R8_5.webp"
        style="max-width: 100%; transition: 0.5s; transform: translateX(0%);"
      />
      <img
        alt="car outside"
        src="/images/AUDI_R8_6.webp"
        style="max-width: 100%; transition: 0.5s; transform: translateX(0%);"
      />
      <img
        alt="car outside"
        src="/images/AUDI_R8_7.webp"
        style="max-width: 100%; transition: 0.5s; transform: translateX(0%);"
      />
      <img
        alt="car outside"
        src="/images/AUDI_R8_8.webp"
        style="max-width: 100%; transition: 0.5s; transform: translateX(0%);"
      />
      <img
        alt="car outside"
        src="/images/AUDI_R8_9.webp"
        style="max-width: 100%; transition: 0.5s; transform: translateX(0%);"
      />
      <img
        alt="car outside"
        src="/images/AUDI_R8_10.webp"
        style="max-width: 100%; transition: 0.5s; transform: translateX(0%);"
      />
      <button
        class="button slider-button right"
        type="button"
      >
        &gt;
      </button>
      <button
        class="button slider-button left"
        disabled=""
        type="button"
      >
        &lt;
      </button>
      <span
        class="span"
      >
        1 / 10
      </span>
      <button
        class="button button "
        type="button"
      >
        <img
          alt="full screen"
          class="img"
          src="full_screen.png"
          style="filter: none;"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
